PythonVisualizations/Heap.py

```diff
@@ -181,13 +184,212 @@ def siftUp(self, i):
+            canvasDimensions[1] - padding,
+            fill=self.OPERATIONS_BG)
+        callEnviron.add(outputBox)
+
+        outputBoxCoords = self.canvas.coords(outputBox)
+        midOutputBox = (outputBoxCoords[3] + outputBoxCoords[1]) // 2
+        # create the value to move to output box
+        valueOutput = self.copyCanvasItem(self.list[peekIndex].display_val)
+        valueList = (valueOutput,)
+        callEnviron.add(valueOutput)
+        # move value to output box
+        # self.highlightCodeTags('print', callEnviron)
+        toPositions = (outputBoxCoords[0] + padding / 2 + (peekIndex + 1 / 2) * spacing,
+                        midOutputBox)
+        self.moveItemsTo(valueList, (toPositions,), sleepTime=.02)
+        # make the value 25% smaller
+        newFont = (self.VALUE_FONT[0], int(self.VALUE_FONT[1] * .75))
+        self.canvas.itemconfig(valueOutput, font=newFont)
+
+        # add label to output box
+        labelX = self.HEAP_X0 * 2 + boxSize * spacing + padding * 5
+        labelY = canvasDimensions[1] - padding * 2
+
+        boxLabel = self.canvas.create_text(
+            labelX, labelY, text="root", font=self.VARIABLE_FONT)
+        callEnviron.add(boxLabel)
+        self.highlightCodeTags([], callEnviron)
+        # finish animation
+        self.cleanUp(callEnviron)
+    
+    def removeMax(self):
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+
+        # remove first element from list
+        n = self.list[0]
+        self.list[0] = None
+        self.size -= 1
+        # if the array is now empty, delete the NoneType place holder from the array
+        if self.size < 1: del self.list[:]
+        # Slide value rectangle up and off screen
+        items = (n.display_shape, n.display_val)
+        toPositions = (100, -100, 100, -50)
+        callEnviron |= set(items)
+        self.moveItemsBy(items, delta=(0, -self.canvas.coords(n.display_shape)[3]), steps=self.CELL_HEIGHT,
+                        sleepTime=.05)
+        self.canvas.delete(items)
+
+        if self.size < 1:
+            self.setMessage("Heap requirement satisfied")
+
+        # move bottom cell to top, and heapify
+        else:
+            self.swapRoot()
+            self.siftDown()
+        
+        # Finish animation
+        self.cleanUp(callEnviron)
+    
+    def swapRoot(self):
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+        n = self.list[-1]
+        items = (n.display_shape, n.display_val)
+
+        # move the last cell to the front of the heap
+        cellCoords = self.cellCoords(0)
+        cellCenter = self.cellCenter(0)
+        toPositions = (cellCoords, cellCenter)
+        # self.moveItemsOffCanvas(items, N, sleepTime=0.02) ## alternative method to moveItemsBy
+        self.moveItemsTo(items, toPositions)
+        self.list[0] = n
+        # delete the last element from the list
+        del self.list[-1]
```

Comment:
I'm concerned that the different representations of the contents of the heap will lead to confusion in making changes to the visualization.  Whenever we make visualizations, we need to have an internal representation of the data along with some extra objects to represent it visually.  It's best to try to keep the internal representation as close as possible to the code that is being visualized.  I'm worried about the use of the `list`, `size`, and `nItems` attributes of the heap being mixed up in this implementation.

The heap is stored in an array.  That array has a capacity or number of cells.  In the `Heap` class, you have defined the `list` as the array, and put the capacity in the `size` attribute.  The `list`'s actual size is being interpreted as `nItems`.   When the textbook code increments `nItems`, the visualization code must append an item to the `list`.  Similarly, decrementing `nItems` means popping off the last item of the `list`.  While it is possible to manage these different representations and keep them consistent, it's easier to make mistakes.

I would suggest keeping the `list` as an array and using its length as the size of the array (instead of the number of active items).  Then instead of a `size` attribute, the `Heap` class maintains a `nItems` attribute to track how many of the cells are in use.  These changes would need to be throughout the Heap.py module, not just in swapRoot().  I point it out here because it changes the capacity of the array on line 369 - `del self.list[-1]` - and there is no reference to changing `nItems`.  Other programmers will need to remember that deleting the last item is being done not to change the capacity, but to decrement `nItems`.   Keeping the visualization structures close to the textbook code makes it easier to see the correspondence of what's on the canvas to the code implementing the algorithm and the code implementing the visualization.

I think another benefit to explicitly representing `nItems` as an attribute of the `Heap` class is that a random fill operation can set the size of the array capacity to the user requested value and set `nItems` to 1.  That means the heap condition is satisfied for all N items.  The Heapify operation, would then use the array capacity to determine how many items are to be heapified.  It would have to skip past any cells at the end (added by the grow heap operation) that do not have a `drawable` record in them when determining the initial value of `heapLo`.

Also, please change `moveItemsTo()` to `moveItemsOnCurve()` in line 366 because it is very hard to see the last item moving into position at the root when it travels along a line through all the other items.  You may have to adjust the `startAngle` parameter based on the distance the value has to move like is done in SimpleSorting.py.